// pages/store/request/requestSuccessModal/requestSuccessModal.js
import React, {useContext} from 'react'
import './requestSuccessModal.css'
import {modalContext} from "contexts/modalContext";

const RequestSuccessModal = () => {

    let {modalContent, handleModal, modal} = useContext(modalContext)

    const closeModal = event => {
        event.preventDefault()
        handleModal()
    }

    const endToEnd = event => {
        event.fontSize = 12
        //this is really cool?
    }

    return (
        <div className="modal modal_ok" style={ModalStyles}>
            <div className="modal_bg"/>
            <div className="modal_box">
                <img onClick={closeModal} className="close mod_close" src="/img/close.svg" alt={"close"}/>
                <img className="ok_img" src="/img/modal_ok_2.svg" alt={"ok_img"}/>
                <h2 style={TitleStyles}>Ваша заявка отправлена</h2>
                <p style={MessageStyles}>
                    {
                        /*<a style={LinkStyles}>Перейти в личный кабинет </a>
                     или
                     */
                    }
                    <a style={LinkStyles}> вернуться к списку автомагазинов</a>
                </p>
            </div>
        </div>

    )
}

export default RequestSuccessModal

const ModalStyles = {
    maxWidth: '508px',
    backgroundColor: '#fff',
    display: 'flex',
    padding: '25px 30px',
    textAlign: 'center',
    position: 'relative'
}

const TitleStyles = {
    color: 'black !important',
    fontSize: '24px',
    lineHeight: '28px',
    margin: '20px 0',
    fontWeight: 'normal'
}

const MessageStyles = {
    fontSize: '14px',
    lineHeight: '16px'
}

const LinkStyles = {
    color: '#FF5254'
}
