// pages/lk/addSubAccount/addAutoparts/addAutoparts.js
import React, {useState, useEffect, useContext} from "react";
import {useForceUpdate} from 'components/useForceUpdate'
import SelectServices from '../selectServices'
import {modalContext} from 'contexts/modalContext'
import useFetch from 'hooks/useFetch'


const AddAutoparts = () => {
    const [images, setImages] = useState([null, null, null, null, null, null])
    //TODO: поставить сюда начальное значение из бд (которого нет)
    const [selectedMainImage, setSelectedMainImage] = useState(null)
    const {modalContent, handleModal, modal} = useContext(modalContext)

    const [{response}, doFetch] = useFetch('/subAccount/create')

    const [selectedServices, setSelectedServices] = useState([])

    const [deliveries, setDeliveries] = useState([])
    const [deliveriesCounter, setDeliveriesCounter] = useState(0)

    const [licenses, setLicenses] = useState([])
    const [licensesConter, setLicensesCounter] = useState(0)

    const [videoReviewLinks, setVideoReviewLinks] = useState([])


    const forceUpdate = useForceUpdate()
    const [vkSoc, setVkSoc] = useState('')
    const [fbSoc, setFbSoc] = useState('')
    const [okSoc, setOkSoc] = useState('')
    const [instSoc, setInstSoc] = useState('')
    const [ytSoc, setYtSoc] = useState('')
    const [twSoc, setTwSoc] = useState('')

    const [title, setTitle] = useState('')
    const [description, setDescription] = useState('')
    const [email, setEmail] = useState('')
    const [street, setStreet] = useState('')
    const [houseNumber, setHouseNumber] = useState('')
    const [phone, setPhone] = useState('')
    const [city, setCity] = useState('')

    useEffect(() => {
        let firstElement = document.getElementById(`mainPhotoCheckbox0`)
        firstElement.checked = true
        setSelectedMainImage(0)
    }, [])

    useEffect(() => {
        console.log(selectedServices)
    }, [selectedServices])

    const handleUploadPhotoLinkClick = event => {
        const inputElement = document.getElementById('uploadPhotoInput')
        inputElement.click()
    }

    const handleCreateButton = event => {
        const servicesIds = []
        selectedServices.forEach(service => {
            servicesIds.push(service.id)
        })
        doFetch({
            method: 'post',
            data: {
                title,
                description,
                category: 'saleAutoParts',
                services: servicesIds,
                address: [{city: 'Уфа', street, houseNumber}],
                phoneNumber: phone,
                workScheduleFrom: 'monday',
                workScheduleTo: 'friday',
                website: '',
                socials: {
                    vk: vkSoc == null ? 'vk.com' : vkSoc,
                    fb: fbSoc == null ? 'facebook.com' : fbSoc,
                    youtube: ytSoc == null ? 'youtube.com' : ytSoc,
                    ok: okSoc == null ? 'ok.com' : okSoc,
                    inst: instSoc == null ? 'instagram.com' : instSoc,
                    tweet: twSoc == null ? 'twitter.com' : twSoc
                }
            }
        })
    }

    //TODO: add check to img and size
    const handleUploadPhotoInput = event => {
        const files = event.target.files
        if (files && files[0]) {
            const img = files[0]

            const reader = new FileReader()
            for (let i = 0; i < images.length; i++) {
                if (images[i] == null) {
                    images[i] = img
                    setImages(images)
                    forceUpdate()
                    const imgTag = document.getElementById(`uploadedPhoto${i}`)
                    imgTag.title = img.name
                    reader.onload = function (event) {
                        imgTag.src = event.target.result
                    }

                    reader.readAsDataURL(img)
                    forceUpdate()
                    break
                }
            }

        }
    }

    const handleDeletePhoto = index => {
        images[index] = null
        setImages(images)
        forceUpdate()
    }

    const handleMainPhotoCheckbox = (event, index) => {
        if (selectedMainImage === null) {
            setSelectedMainImage(index)
            return
        }

        if (index === selectedMainImage) {
            console.log('prevSelected')
            event.preventDefault()
            return
        }

        let prevSelectedElement = document.getElementById(`mainPhotoCheckbox${selectedMainImage}`)
        prevSelectedElement.checked = false
        console.log(prevSelectedElement)
        setSelectedMainImage(index)
    }

    const handleClickOnService = (event, id) => {
        let updatedItems = selectedServices
        const ids = updatedItems.map(el => el.id)
        const index = ids.indexOf(id)

        updatedItems.splice(index, 1)
        setSelectedServices(updatedItems)
        forceUpdate()
    }

    const handleCitySelectClick = event => {

    }

    const handleSelectAutoMarks = event => {
        event.preventDefault()
        handleModal(
            <SelectServices
                selectedServices={selectedServices}
                setSelectedServices={setSelectedServices}
            />
        )
    }

    const handleLoadDelivery = event => {
        const inputElement = document.getElementById('uploadDeliveryInput')
        inputElement.click()
    }

    const handleUploadDeliveryInput = event => {
        const files = event.target.files
        if (!files || !files[0]) {
            return;
        }

        const selectedImage = files[0]
        const reader = new FileReader()
        deliveries.push(selectedImage)
        setDeliveriesCounter(deliveriesCounter + 1)
        setDeliveries(deliveries)
        forceUpdate()

        setTimeout(() => {
            const imgTag = document.getElementById(`deliveryPhoto${deliveriesCounter}`)
            imgTag.title = selectedImage.name

            reader.onload = function (event) {
                imgTag.src = event.target.result
            }

            reader.readAsDataURL(selectedImage)
            forceUpdate()
        }, 1000)
    }
    const handleUploadLicenseInput = (event) => {
        const files = event.target.files
        if (!files || !files[0]) {
            return;
        }

        const selectedImage = files[0]
        const reader = new FileReader()
        licenses.push(selectedImage)
        setLicensesCounter(licensesConter + 1)
        setLicenses(licenses)
        forceUpdate()

        setTimeout(() => {
            const imgTag = document.getElementById(`licensePhoto${licensesConter}`)
            imgTag.title = selectedImage.name

            reader.onload = function (event) {
                imgTag.src = event.target.result
            }

            reader.readAsDataURL(selectedImage)
            forceUpdate()
        }, 1000)
        forceUpdate()
    }
    const handleLoadLicenses = event => {
        const inputElement = document.getElementById('uploadLicenseInput')
        inputElement.click()
    }
    const handleLoadVideoReview = event => {
        const el = document.getElementById('videoReviewInput')
        let ytLink = el.value

        //add check is youtube link in input
        if (!ytLink.includes('/embed/')) {
            const arr = ytLink.split('watch?v=')
            ytLink = `${arr[0]}/embed/${arr[1]}`
        }

        videoReviewLinks.push(ytLink)
        setVideoReviewLinks(videoReviewLinks)
        el.value = ''
        forceUpdate()
    }

    const handleVideoClose = (index) => {
        videoReviewLinks.splice(index, 1)
        forceUpdate()
    }




    return (

        <section className="mag_block">
            <div className="container">
                <h1 className="title5">Информация о магазине</h1>
                <div className="mag_info">
                    <div className="mag_name">
                        <label>Название компании</label>
                        <input
                            value={title}
                            onChange={e => setTitle(e.target.value)}
                            minLength={3}
                            maxLength={25}
                            type="text"
                            placeholder="Введите название"
                        />
                    </div>
                    <div className="mag_text_foto">
                        <div className="mag_text_foto_item">
                            <label>Описание компании</label>
                            <textarea
                                maxLength={700}
                                minLength={10}
                                value={description}
                                placeholder="Расскажите котротко о своем магазине"
                                onChange={e => setDescription(e.target.value)}
                            />
                            <p>Максимальная количество текста: до 700 знаков с пробелами</p>
                        </div>
                        <div className="mag_text_foto_item">

                            <label>Фотографии</label>
                            <div className="mag_foto">
                                <div className="mag_foto_load">
                                    <div style={{cursor: 'pointer'}} id={'uploadPhotoLink'}
                                         onClick={handleUploadPhotoLinkClick}
                                         className="mag_foto_load_img">
                                        <img src="/img/fotoapp.svg"/>
                                        <p>Загрузить фото компании</p>
                                    </div>
                                    <p>Загрузите до 6 фотографий, общий объем не должен превышать 50 мб.</p>
                                </div>
                                <input
                                    id={'uploadPhotoInput'}
                                    onChange={handleUploadPhotoInput}
                                    type={'file'}
                                    style={{display: 'none'}}
                                />
                                <div className="mag_foto_gallery">
                                    {
                                        images.map((image, index) => (
                                            <div key={index} className="mag_foto_gallery_item">
                                                {
                                                    images[index] != null ? (
                                                        <div className="mag_foto_gallery_none"
                                                             style={{display: 'none'}}>
                                                            <img src="/img/clarity_image-line.svg"
                                                                 />
                                                            <p>Нет фото</p>
                                                        </div>
                                                    ) : (
                                                        <div className="mag_foto_gallery_none">
                                                            <img src="/img/clarity_image-line.svg"
                                                                />
                                                            <p>Нет фото</p>
                                                        </div>
                                                    )
                                                }

                                                {
                                                    images[index] == null ? (
                                                        <div
                                                            style={{display: 'none'}}
                                                            className="mag_foto_gallery_img">
                                                            <img className="mag_foto_gallery_foto"
                                                                 id={`uploadedPhoto${index}`}
                                                                 src={'/img/test10.jpg'}/>
                                                            <img className="mag_foto_gallery_close"
                                                                 src="/img/close_foto.svg"
                                                                 />
                                                        </div>
                                                    ) : (
                                                        <div
                                                            className="mag_foto_gallery_img">
                                                            <img className="mag_foto_gallery_foto"
                                                                 id={`uploadedPhoto${index}`}
                                                                 src={'/img/test10.jpg'} />
                                                            <img onClick={e => handleDeletePhoto(index)}
                                                                 className="mag_foto_gallery_close"
                                                                 src="/img/close_foto.svg"
                                                                />
                                                        </div>
                                                    )
                                                }

                                                <div className="mag_foto_gallery_big">
                                                    <label>
                                                        <input
                                                            onClick={e => handleMainPhotoCheckbox(e, index)}
                                                            type="checkbox"
                                                            id={`mainPhotoCheckbox${index}`}
                                                        />
                                                        <div>Главное</div>
                                                    </label>
                                                </div>
                                            </div>
                                        ))
                                    }


                                </div>
                            </div>
                        </div>
                    </div>
                    <div className="mag_mark">
                        <h2 className="title5 up_90">Выберите марки автомобилей запчасти для которых Вы продаете</h2>
                        <div className="mag_mark_row">
                            <a onClick={handleSelectAutoMarks} className="but1">
                                Выбрать марки
                                <img src="/img/plus_red.svg" />
                            </a>
                            {selectedServices.length === 0 ? (
                                <p className="none_cont">Вы пока не выбрали ни одну марку</p>
                            ) : (
                                selectedServices.map((selected, index) => (
                                    <div onClick={event => {
                                        handleClickOnService(event, selected.id)
                                    }} key={index} className="mag_mark_item">
                                        {selected.title}
                                    </div>
                                ))
                            )}
                        </div>
                    </div>
                    <div className="mag_comp">
                        <h2 className="title5 up_9">Транспортные компании с которыми Вы работаете</h2>
                        <div className="mag_comp_row" id={'magCompRowDelivery'}>
                            <div className="mag_comp_load" id={'loadDelivery'} id={'magCompLoadDelivery'}>
                                <div onClick={handleLoadDelivery} className="mag_comp_load_box"
                                     style={{cursor: 'pointer'}}>
                                    <img src="/img/big_plus_red.svg" />
                                    <p>Загрузить фото/лого транспортной компании</p>
                                </div>

                                <input
                                    id={'uploadDeliveryInput'}
                                    onChange={handleUploadDeliveryInput}
                                    type={'file'}
                                    style={{display: 'none'}}
                                />
                            </div>


                            <div className="mag_comp_item">
                                <img src="/img/test13.png"/>
                                <button/>
                            </div>

                            {
                                deliveries.map((delivery, index) => (
                                    <div key={index} className="mag_comp_item">
                                        <img
                                            id={`deliveryPhoto${index}`}
                                            src={'/img/test13.jpg'}
                                        />
                                    </div>
                                ))
                            }
                        </div>
                    </div>
                    <div className="mag_video">
                        <h2 className="title5 up_90">Видеоотзывы</h2>
                        <div className="mag_video_row">
                            <div className="mag_video_additem">
                                <div className="mag_video_additem_box">
                                    <h3>Укажите ссылку на видеоотзыв,<br/> предварительно разместив его на Youtube</h3>
                                    <input type="text" id={'videoReviewInput'} placeholder="https://www.youtube.com"/>
                                    <a onClick={handleLoadVideoReview} className="but1">
                                        Добавить
                                        <img src="/img/plus_red.svg" />
                                    </a>
                                </div>
                            </div>
                            <p className="none_cont" style={{display: 'none'}}>Вы пока не выбрали ни одну марку</p>
                            {
                                videoReviewLinks.length !== 0 && videoReviewLinks.map((videoReview, index) => (
                                    <div className="mag_video_item" key={index}>
                                        <iframe
                                            width="300"
                                            height="200"
                                            src={videoReview}
                                            frameBorder="0"
                                            allowFullScreen
                                        />
                                        <button onClick={e => {
                                            handleVideoClose(index)
                                        }}/>
                                    </div>
                                ))
                            }
                        </div>
                    </div>
                    <div className="mag_lic">
                        <h2 className="title5 up_90">Лицензии, сертификаты, награды</h2>
                        <div className="mag_comp_row">
                                <div className="mag_comp_load" id={'loadLicense'}>
                                    <div onClick={handleLoadLicenses} className="mag_comp_load_box"
                                         style={{cursor: 'pointer'}}>
                                        <img src="/img/big_plus_red.svg" />
                                        <p>Загрузить фото лицензии или сертификата
                                        </p>
                                    </div>

                                    <input
                                        id={'uploadLicenseInput'}
                                        onChange={handleUploadLicenseInput}
                                        type={'file'}
                                        style={{display: 'none'}}
                                    />
                                </div>
                                {
                                    licenses.map((license, index) => (
                                        <div key={index} className="mag_lic_item mag_comp_load">
                                            <img
                                                id={`licensePhoto${index}`}
                                                src={'/img/test13.jpg'}
                                            />
                                        </div>
                                    ))
                                }
                            </div>
                    </div>
                    <div className="mag_seti">
                        <h2 className="title5 up_90">Сылки на Ваши соц. сети</h2>
                        <div className="mag_seti_row">
                            <div className="mag_seti_item">
                                <input
                                    value={vkSoc}
                                    onChange={e => setVkSoc(e.target.value)}
                                    type="text"
                                    className="vk"
                                />
                            </div>
                            <div className="mag_seti_item">
                                <input
                                    value={fbSoc}
                                    onChange={e => setFbSoc(e.target.value)}
                                    type="text"
                                    className="fb"
                                />
                            </div>
                            <div className="mag_seti_item">
                                <input
                                    value={okSoc}
                                    onChange={e => setOkSoc(e.target.value)}
                                    type="text"
                                    className="ok"
                                />
                            </div>
                            <div className="mag_seti_item">
                                <input
                                    value={twSoc}
                                    onChange={e => setTwSoc(e.target.value)}
                                    type="text"
                                    className="tw"
                                />
                            </div>
                            <div className="mag_seti_item">
                                <input
                                    value={instSoc}
                                    onChange={e => setInstSoc(e.target.value)}
                                    type="text"
                                    className="inst"
                                />
                            </div>
                            <div className="mag_seti_item">
                                <input
                                    value={ytSoc}
                                    onChange={e => setYtSoc(e.target.value)}
                                    type="text"
                                    className="yt"
                                />
                            </div>
                        </div>
                    </div>
                </div>
                <div className="mag_adr">
                    <h2 className="title5">
                        Адрес и время работы
                    </h2>
                    <div className="mag_adr_inf">
                        <div className="mag_adr_town active">
                            <label>Город</label>
                            <div onClick={handleCitySelectClick} className="mag_adr_town_box">
                                <p><b>Уфа</b></p>
                            </div>
                        </div>
                        <div className="mag_adr_streat">
                            <label>Улица</label>
                            <input
                                value={street}
                                onChange={e => setStreet(e.target.value)}
                                type="text"
                                placeholder="Улица: Пушкина"
                            />
                        </div>
                        <div className="mag_adr_dom">
                            <label>Дом</label>
                            <input
                                value={houseNumber}
                                onChange={e => setHouseNumber(e.target.value)}
                                type="text"
                                placeholder="49"
                            />
                        </div>
                    </div>
                    <div className="mag_adr_box">
                        <div className="mag_adr_cont" style={{height: 300}}>
                            <div className="mag_adr_cont_up" style={{marginBottom: -100}}>
                                <div className="mag_adr_cont_tel">
                                    <label>Телефон</label>
                                    <div className="mag_adr_cont_tel_row">
                                        <div className="mag_adr_cont_tel_item">
                                            <input
                                                value={phone}
                                                onChange={e => setPhone(e.target.value)}
                                                maxLength={11}
                                                minLength={11}
                                                type="tel"
                                                placeholder="+ 7 (123) 345 67 89"
                                            />
                                        </div>
                                    </div>
                                </div>
                            </div>
                            <div className="mag_adr_cont_down" style={{marginTop: -100}}>
                                <div className="mag_adr_cont_email">
                                    <label>E-mail*</label>
                                    <input
                                        type="email"
                                        value={email}
                                        onChange={e => setEmail(e.target.value)}
                                        placeholder="info@yandex.ru"
                                    />
                                </div>
                                <div className="mag_adr_cont_time_date">
                                    <div className="mag_adr_cont_time">
                                        <label>График работы</label>
                                        <input
                                            type="text"
                                            placeholder="Пн - Пт"
                                        />
                                    </div>
                                    <div className="mag_adr_cont_date">
                                        <label>Время работы</label>
                                        <input
                                            type="text"
                                            placeholder="09:00 - 18:00"
                                        />
                                    </div>
                                </div>
                            </div>
                        </div>
                        <div className="mag_adr_map">
                            <h3>Месторасположение на карте</h3>
                            <div className="mag_adr_map_box">
                                <iframe className="map1"
                                        src="https://yandex.ru/map-widget/v1/?um=constructor%3Af46f0c3a6fc272b0566fb4d250bb7e47e320d8355838f8bab5242008049d1544&amp;source=constructor"
                                        width="100%" height="405" frameBorder="0"/>
                            </div>
                        </div>
                    </div>
                    <a onClick={handleCreateButton} className="mag_adr_add_map but1">Создать</a>
                </div>
            </div>
        </section>
    )
}

export default AddAutoparts
